### PR TITLE
GC docs

### PR DESCRIPTION
**Pull request by d-netto**
Extends the documentation work from https://github.com/JuliaLang/julia/pull/45246.

### CODE REVIEW

File: doc/src/devdocs/gc.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+- Object pools, malloc'd arrays and big objects are sweeped. On a full sweep, the mark bits of all marked objects are reset. On a generational sweep,
+only the mark bits of marked young objects are reset.
+
+- Mark bits of objects in the remembered set are set, so we don't trigger the write barrier on them again.
+
+After these stages, old objects will be left with their mark bits set, so that references from them are not explored in a subsequent generational collection. This scheme eliminates the need of explicitly keeping a flag to indicate a full mark (though a flag to indicate a full sweep is necessary).
+
+## Heuristics
+
+GC heuristics tune the GC by changing the size of the allocation interval between garbage collections. If a GC was unproductive than we increase the size of the allocation interval to allow objects more time to die. If a GC returns a lot of space we can shrink the interval. The goal is to find a steady state where we are allocating just about the same amount as we are collecting.
```
Comment (matthias314) [inline doc/src/devdocs/gc.md:50]: typo:
```
If a GC was unproductive, then ...
```

**Comment by ViralBShah** [comment]
cc @chflood

**Comment by d-netto** [comment]
Bump?

**Comment by oscardssmith** [comment]
this is good to merge as soon as CI passes.

**Comment by KristofferC** [comment]
Doesn't this need to be referred to from somewhere in order to show up in the manual?

File: doc/src/devdocs/gc.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+one for sufficiently large malloc'd arrays (`mallocarray_t`) and one for
+sufficiently large objects (`bigval_t`).
+
+Sweeping these objects consists of unlinking them from their list and calling `free` on the
+corresponding address.
+
+## Generational and remembered sets
+
+Field writes into old objects trigger a write barrier if the written field
+points to a young object and if a write barrier has not been triggered in the old object yet.
```
Comment (vchuravy) [inline doc/src/devdocs/gc.md:38]:
```suggestion
points to a young object and if a write barrier has not been triggered on the old object yet.
```

File: doc/src/devdocs/gc.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+Julia has a generational non-moving mark-sweep garbage collector.
+Native objects are precisely scanned and foreign ones are conservatively marked.
+
+## Memory layout of objects and GC bits
+
+An opaque tag is stored in the front of GC managed objects, and its lowest two bits are
+used for garbage collection.  The lowest bit is set for marked objects and the second
+lowest bit stores age information (e.g. it's only set for old objects).
+
+Objects are allocated in multiple of 4 bytes to ensure this pointer tagging is legal.
```
Comment (vchuravy) [inline doc/src/devdocs/gc.md:14]:
```suggestion
Objects are aligned by a multiple of 4 bytes to ensure this pointer tagging is legal.
```

**Review by vchuravy — APPROVED** [review]

**Pull request merged**